datumaro/components/extractor.py

```diff
@@ -624,10 +624,10 @@ def _find_sources_recursive(cls, path, ext, extractor_name,
             for d in range(max_depth + 1):
                 sources.extend({'url': p, 'format': extractor_name} for p in
                     iglob(osp.join(path, *('*' * d), dirname, filename + ext)))
+                if callable(file_filter):
```

Comment:
This way the function will be called on each iteration, and will re-check items from previous steps. We can move it out of the loop or include inside of `extend(...)`.